File: lib/core/design_system/widgets/fc_button.dart
import 'package:flutter/material.dart';

import '../fc_colors.dart';
import '../fc_theme.dart';
import '../fc_tokens.dart';

/// Visual variants — 1:1 with `components/ui/button.tsx`'s `buttonVariants`.
enum FCButtonVariant { primary, destructive, outline, secondary, ghost, link }

/// Sizes — 1:1 with `buttonVariants.size`. (`primary` == cva `default`.)
enum FCButtonSize { md, sm, lg, icon }

/// A custom button matching shadcn/ui's `<Button>` — no Material elevation or
/// ripple. Hover states from the web (`hover:bg-primary/90`, `/80`) are mapped
/// to the pressed state for touch via alpha-composited colours.
class FCButton extends StatefulWidget {
  const FCButton({
    super.key,
    this.label,
    this.onPressed,
    this.variant = FCButtonVariant.primary,
    this.size = FCButtonSize.md,
    this.icon,
    this.trailingIcon,
    this.fullWidth = false,
    this.loading = false,
  }) : assert(label != null || icon != null,
            'FCButton needs a label or an icon');

  final String? label;
  final VoidCallback? onPressed;
  final FCButtonVariant variant;
  final FCButtonSize size;

  /// Leading icon (rendered at 16px, matching `[&_svg]:size-4`).
  final Widget? icon;

  /// Trailing icon (16px).
  final Widget? trailingIcon;

  /// Stretch to the parent's width (the web `w-full`).
  final bool fullWidth;

  /// Shows a spinner and disables interaction.
  final bool loading;

  bool get _enabled => onPressed != null && !loading;

  @override
  State<FCButton> createState() => _FCButtonState();
}

class _FCButtonState extends State<FCButton> {
  bool _pressed = false;

  void _setPressed(bool value) {
    if (_pressed != value) setState(() => _pressed = value);
  }

  @override
  Widget build(BuildContext context) {
    final c = context.colors;
    final style = _resolveStyle(c, _pressed && widget._enabled);
    final metrics = _sizeMetrics(widget.size);

    final textStyle = TextStyle(
      fontSize: FCFontSizes.sm,
      fontWeight: FCFontWeights.medium,
      color: style.foreground,
      decoration:
          widget.variant == FCButtonVariant.link && _pressed && widget._enabled
              ? TextDecoration.underline
              : TextDecoration.none,
      decorationColor: style.foreground,
      height: 1.0,
    );

    Widget content;
    if (widget.loading) {
      content = SizedBox(
        width: 16,
        height: 16,
        child: CircularProgressIndicator(
          strokeWidth: 2,
          valueColor: AlwaysStoppedAnimation<Color>(style.foreground),
        ),
      );
    } else {
      final children = <Widget>[
        if (widget.icon != null) widget.icon!,
        if (widget.label != null) Flexible(child: Text(widget.label!, maxLines: 1, overflow: TextOverflow.ellipsis)),
        if (widget.trailingIcon != null) widget.trailingIcon!,
      ];
      content = Row(
        mainAxisSize: widget.fullWidth ? MainAxisSize.max : MainAxisSize.min,
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          for (var i = 0; i < children.length; i++) ...[
            if (i > 0) const SizedBox(width: 8), // gap-2
            children[i],
          ],
        ],
      );
    }

    final button = AnimatedContainer(
      duration: FCDurations.transition,
      curve: Curves.easeOut,
      height: metrics.height,
      width: widget.size == FCButtonSize.icon
          ? metrics.height
          : (widget.fullWidth ? double.infinity : null),
      padding: widget.size == FCButtonSize.icon
          ? EdgeInsets.zero
          : EdgeInsets.symmetric(horizontal: metrics.horizontalPadding),
      decoration: BoxDecoration(
        color: style.background,
        borderRadius: FCRadii.mdAll,
        border: style.border != null
            ? Border.all(color: style.border!, width: 1)
            : null,
      ),
      child: Center(
        widthFactor: widget.fullWidth ? null : 1,
        child: IconTheme.merge(
          data: IconThemeData(color: style.foreground, size: 16),
          child: DefaultTextStyle.merge(style: textStyle, child: content),
        ),
      ),
    );

    return Opacity(
      opacity: widget._enabled ? 1.0 : 0.5,
      child: Semantics(
        button: true,
        enabled: widget._enabled,
        label: widget.label,
        child: GestureDetector(
          behavior: HitTestBehavior.opaque,
          onTap: widget._enabled ? widget.onPressed : null,
          onTapDown: widget._enabled ? (_) => _setPressed(true) : null,
          onTapUp: widget._enabled ? (_) => _setPressed(false) : null,
          onTapCancel: widget._enabled ? () => _setPressed(false) : null,
          child: MouseRegion(
            cursor: widget._enabled
                ? SystemMouseCursors.click
                : SystemMouseCursors.basic,
            child: button,
          ),
        ),
      ),
    );
  }

  _ButtonStyle _resolveStyle(FCColors c, bool pressed) {
    switch (widget.variant) {
      case FCButtonVariant.primary:
        // default: bg-primary text-primary-foreground hover:bg-primary/90
        return _ButtonStyle(
          background: pressed ? c.primary.withValues(alpha: 0.9) : c.primary,
          foreground: c.primaryForeground,
        );
      case FCButtonVariant.destructive:
        return _ButtonStyle(
          background:
              pressed ? c.destructive.withValues(alpha: 0.9) : c.destructive,
          foreground: c.destructiveForeground,
        );
      case FCButtonVariant.outline:
        // border border-input bg-background hover:bg-accent hover:text-accent-foreground
        return _ButtonStyle(
          background: pressed ? c.accent : c.background,
          foreground: pressed ? c.accentForeground : c.foreground,
          border: c.input,
        );
      case FCButtonVariant.secondary:
        return _ButtonStyle(
          background:
              pressed ? c.secondary.withValues(alpha: 0.8) : c.secondary,
          foreground: c.secondaryForeground,
        );
      case FCButtonVariant.ghost:
        return _ButtonStyle(
          background: pressed ? c.accent : Colors.transparent,
          foreground: pressed ? c.accentForeground : c.foreground,
        );
      case FCButtonVariant.link:
        return _ButtonStyle(
          background: Colors.transparent,
          foreground: c.primary,
        );
    }
  }

  _SizeMetrics _sizeMetrics(FCButtonSize size) {
    switch (size) {
      case FCButtonSize.md:
        return const _SizeMetrics(height: 40, horizontalPadding: 16);
      case FCButtonSize.sm:
        return const _SizeMetrics(height: 36, horizontalPadding: 12);
      case FCButtonSize.lg:
        return const _SizeMetrics(height: 44, horizontalPadding: 32);
      case FCButtonSize.icon:
        return const _SizeMetrics(height: 40, horizontalPadding: 0);
    }
  }
}

class _ButtonStyle {
  const _ButtonStyle({
    required this.background,
    required this.foreground,
    this.border,
  });

  final Color background;
  final Color foreground;
  final Color? border;
}

class _SizeMetrics {
  const _SizeMetrics({required this.height, required this.horizontalPadding});

  final double height;
  final double horizontalPadding;
}

File: app/src/main/kotlin/com/finnacalc/android/core/designsystem/FCButton.kt
//
// FCButton.kt
//
// Port of iOS FCButton.swift (itself a native port of the web
// components/ui/button.tsx). Web `hover:` states map to the pressed state
// (the /90, /80 opacity steps); `disabled:opacity-50` maps to `enabled`.
//

package com.finnacalc.android.core.designsystem

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.ProvideTextStyle
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp

/** Mirrors buttonVariants.variant in button.tsx. */
enum class FCButtonVariant {
    Default,      // bg-primary text-primary-foreground hover:bg-primary/90
    Destructive,  // bg-destructive text-destructive-foreground hover:bg-destructive/90
    Outline,      // border border-input bg-background hover:bg-accent
    Secondary,    // bg-secondary text-secondary-foreground hover:bg-secondary/80
    Ghost,        // hover:bg-accent hover:text-accent-foreground
    Link,         // text-primary underline on press
}

/** Mirrors buttonVariants.size in button.tsx. */
enum class FCButtonSize {
    Default,  // h-10 px-4
    Sm,       // h-9  px-3
    Lg,       // h-11 px-8
    Icon,     // h-10 w-10
}

/**
 * A button styled to match the web `<Button>`. Slot content for arbitrary
 * label (icon + text); use the String overload for plain text.
 */
@Composable
fun FCButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    variant: FCButtonVariant = FCButtonVariant.Default,
    size: FCButtonSize = FCButtonSize.Default,
    enabled: Boolean = true,
    content: @Composable () -> Unit,
) {
    val c = Theme.colors
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()

    val height = when (size) {
        FCButtonSize.Default, FCButtonSize.Icon -> 40.dp  // h-10
        FCButtonSize.Sm -> 36.dp                          // h-9
        FCButtonSize.Lg -> 44.dp                          // h-11
    }
    val hPad = when (size) {
        FCButtonSize.Default -> 16.dp  // px-4
        FCButtonSize.Sm -> 12.dp       // px-3
        FCButtonSize.Lg -> 32.dp       // px-8
        FCButtonSize.Icon -> 0.dp      // square
    }

    // Background; `pressed` applies the web hover: opacity step.
    val background = when (variant) {
        FCButtonVariant.Default -> c.primary.copy(alpha = if (pressed) 0.9f else 1f)
        FCButtonVariant.Destructive -> c.destructive.copy(alpha = if (pressed) 0.9f else 1f)
        FCButtonVariant.Secondary -> c.secondary.copy(alpha = if (pressed) 0.8f else 1f)
        FCButtonVariant.Outline -> if (pressed) c.accent else c.background
        FCButtonVariant.Ghost -> if (pressed) c.accent else Color.Transparent
        FCButtonVariant.Link -> Color.Transparent
    }
    val foreground = when (variant) {
        FCButtonVariant.Default -> c.primaryForeground
        FCButtonVariant.Destructive -> c.destructiveForeground
        FCButtonVariant.Secondary -> c.secondaryForeground
        FCButtonVariant.Outline, FCButtonVariant.Ghost ->
            if (pressed) c.accentForeground else c.foreground
        FCButtonVariant.Link -> c.primary
    }
    val animatedBackground by animateColorAsState(
        background, tween(150), label = "fcButtonBackground" // transition-colors
    )

    val shape = RoundedCornerShape(Theme.Radius.md)
    Row(
        modifier = modifier
            .height(height)
            .then(if (size == FCButtonSize.Icon) Modifier.width(40.dp) else Modifier)
            .alpha(if (enabled) 1f else 0.5f)  // disabled:opacity-50
            .clip(shape)
            .background(animatedBackground)
            .then(
                if (variant == FCButtonVariant.Outline) {
                    Modifier.border(1.dp, c.input, shape)  // border border-input
                } else Modifier
            )
            .clickable(
                interactionSource = interaction,
                indication = null,  // press feedback is the color step, as on iOS
                enabled = enabled,
                onClick = onClick,
            )
            .padding(horizontal = hPad),
        horizontalArrangement = Arrangement.spacedBy(8.dp, Alignment.CenterHorizontally),  // gap-2
        verticalAlignment = Alignment.CenterVertically,
    ) {
        CompositionLocalProvider(LocalContentColor provides foreground) {
            ProvideTextStyle(
                Theme.sans(Theme.FontSize.sm, FontWeight.Medium).copy(
                    color = foreground,
                    textDecoration = if (variant == FCButtonVariant.Link && pressed) {
                        TextDecoration.Underline
                    } else TextDecoration.None,
                )
            ) {
                content()
            }
        }
    }
}

/** Convenience for a plain text button: `FCButton("Save") { ... }`. */
@Composable
fun FCButton(
    title: String,
    modifier: Modifier = Modifier,
    variant: FCButtonVariant = FCButtonVariant.Default,
    size: FCButtonSize = FCButtonSize.Default,
    enabled: Boolean = true,
    onClick: () -> Unit,
) {
    FCButton(onClick, modifier, variant, size, enabled) { Text(title) }
}
